Feat/update bound actor

JacobBumgarner opened the pull request:
### Overview
Sometimes when plotting, the physical bounds of the plotter don't match the values of the values shown on the bounds axis bar. For example, when plotting datasets in [VesselVio](https://github.com/jacobbumgarner/vesselvio/), the vasculature from all datasets is normalized prior to plotting. This means that when calling plotter.show_bounds(), the scale of the bounds axis will be incorrect if the input resolution wasn't 1.

@adeak and I had a brief discussion about this, and he found that the vtkCubeAxesActor can indeed take specific calls to separate the physical bounds of the mesh from the numerical bounds shown on the axes.


### Details

To implement this, an optional bounds_range argument was added to the def show_bounds() function in the Renderer class from the renderer.py module.

As seen in the file change, the arg takes a [xmin, xmax, ymin, ymax, zmin, zmax] list/tuple and sets the values on the axes accordingly.

Because the semantics of the plotter bounds vs. the axis actor bounds might be tricky, I decided to use the range vocabulary based on the functions that are called on the vtkCubeAxesActor ([see doc](https://vtk.org/doc/nightly/html/classvtkCubeAxesActor.html)). Not sure if it's perfect, but it's a starting point!

adeak commented:
> Does it make sense to add a test for this functionality? I'm not sure if we can verify that the bounds are changed, but at least to verify that the code is runnable with the expected inputs.

There seem to be getters as well as setters for the bounds (e.g. https://vtk.org/doc/nightly/html/classvtkCubeAxesActor.html#a190c240116715378fb00d7749e913b1e), so querying with the getter after setting in the python API would make sense. Then again perhaps we could/should expose the bounds too as a property or method?

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1273,17 @@ def show_bounds(
         cube_axes_actor.SetBounds(bounds)
 
+        # update the values of range of each axis
+        if isinstance(bounds_range, (list, tuple)):
```
I'd consider being more liberal here. `collections.abc.Sequence` for one, but that doesn't include numpy ndarrays which might also appear here.

banesullivan replied:
There are quite a few occurrences of `isinstance(..., (list, tuple))` we may want to address

adeak replied:
@banesullivan that should probably be a separate PR.

JacobBumgarner replied:
@adeak I agree. I can imagine the `isinstance()`growing quite large though. 

Any thoughts on an approach that can elegantly deal with this other than just adding to the tuple?

```python
if isinstance(..., (list, tuple, 
                    np.ndarray, collections.abc.Sequence)):
```

adeak replied:
`collections.abc.Sequence` includes stdlib sequences, so lists and tuples too.

banesullivan reviewed (dismissed):
This looks nice, @JacobBumgarner! Would you be able to add a trivial test for code coverage?

Also, I'm curious if you could provide a screenshot (perhaps two subplots) showing this feature and how it improves the bounds actor for the types of data you mentioned?

JacobBumgarner commented:
Thanks @banesullivan! Yes - I will add a few example images showing exactly how the bounds actor changes with this feature (long week, will get to it this weekend ;)

And I am more than happy to write a test for code coverage, but I'm afraid to say that I haven't had to do that before. If you could point me to some guidance doc, I will take care of it!

akaszynski commented:
> And I am more than happy to write a test for code coverage, but I'm afraid to say that I haven't had to do that before. If you could point me to some guidance doc, I will take care of it!

Take a look at [test_renderer.py](https://github.com/pyvista/pyvista/blob/main/tests/test_renderer.py) and add a unit test there. You'll want to add a simple function that tests the functionality of `Renderer.show_bounds`. Be sure to add in some [`assert` statements](https://docs.python.org/3/reference/simple_stmts.html), ensuring that you've changed the values of the underlying VTK render. Run your tests locally with `pytest tests/test_renderer.py` and be sure they pass.

JacobBumgarner commented:
Ok - update that should hopefully fix the requests above.
**General Change:** I've changed the bounds_range to axes_range. I imagine disambiguating the difference between `bounds` and `bounds_range` would probably be more difficult than the difference between `bounds` and `axes_range`. To me, `axes_range` seems more pythonic and in the essence of PyVista. Happy to change it back if others disagree. 

---

@banesullivan 
Here is an example screenshot that demonstrates how sending an `axes_range` at a scale argument alters the plotter. The left mesh has a 1 unit^3 resolution, and the right mesh has a 2 unit^3 resolution, i.e., the XYZ axes ranges are all 2x larger.
<p align="center">
<img src="https://user-images.githubusercontent.com/70919881/162584293-7040dd50-9f11-469b-82bc-5b10187694ed.png" width="300" height="300"/>
</p>


<details>
<summary>Dropdown for the example code. Could be simplified by just adding a sphere.</summary>

```python
import numpy as np
import pyvista as pv

plotter = pv.Plotter(shape=(1,2), window_size=(1000, 1000), 
                     multi_samples=4, off_screen=True)

def create_tube():
    theta = np.linspace(-4*np.pi, 4*np.pi, 100, endpoint=True)
    x = np.cos(theta)
    y = np.sin(theta)
    z = np.linspace(-4, 4, 100, endpoint=True)
    tube_coords = np.vstack([x, y, z]).T

    # Example tube radius variation
    r = np.log10(np.linspace(2, 10, 50)) / 2
    tube_radius = np.append(r, np.flip(r))
    radius_factor = np.max(tube_radius) / np.min(tube_radius)

    # Create tube
    spline = pv.Spline(tube_coords)
    spline['radius'] = tube_radius
    tube = spline.tube(radius=np.min(tube_radius), scalars='radius',
                    radius_factor=radius_factor)
    return tube

tube = create_tube()

# Add mesh to first plotter.
plotter.subplot(0, 0)
plotter.add_mesh(tube, smooth_shading=True)
# Mesh loaded at a 1 unit^3 resolution, meaning the axes_range = mesh.bounds
plotter.show_bounds(location='front') 

# Add mesh to second plotter.
plotter.subplot(0, 1)
plotter.add_mesh(tube, smooth_shading=True)
# Mesh loaded at a 2 unit^3 resolution, meaning the axes_range = mesh.bounds * 2
axes_range = np.array(tube.bounds) * 2
plotter.show_bounds(axes_range=axes_range, location='front')

# Show the plotter

# plotter.show()
plotter.screenshot('demo.png')
```
</details>

---

@adeak Thanks for educating - `(tuple, list, collections.abc.Sequence, np.ndarray)` was replaced with `(collections.abc.Sequence, np.ndarray)` 

---

@akaszynski I've updated the `tests/test_renderer.py` file. 

Thanks all!

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1249,6 +1255,9 @@ def show_bounds(
         # set bounds
         if bounds is None:
             bounds = np.array(mesh.GetBounds())
```
I wonder why this didn't do this instead:
```suggestion
            bounds = np.array(mesh.bounds)
```

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1070,6 +1071,11 @@ def show_bounds(
             Bounds to override mesh bounds in the form ``[xmin, xmax,
             ymin, ymax, zmin, zmax]``.
 
+        axes_range : list, tuple, or np.ndarray, optional
```
+1 for the new name.

JacobBumgarner replied:
I'm also thinking that it should be `axes_ranges`.

JacobBumgarner replied:
Thoughts on my updated implementation of the `plotter.axes_ranges` `@property`? I'm probably driving you all nuts with this minor feature implementation lol

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1249,6 +1255,9 @@ def show_bounds(
         if bounds is None:
             bounds = np.array(mesh.GetBounds())
+        else:
+            bounds = np.array(bounds, dtype=np.float_)
```
If `bounds` is already a float64-dtype array (or array subclass) then this will make this line a no-op:
```suggestion
            bounds = np.asanyarray(bounds, dtype=np.float_)
```
I'd also be tempted to just say `dtype=float` because that's shorter, but I guess being explicit is fine too (the end result is the same).

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1249,6 +1255,9 @@ def show_bounds(
+            bounds = np.array(bounds, dtype=np.float_)
+
         if isinstance(padding, (int, float)) and 0.0 <= padding < 1.0:
             if not np.any(np.abs(bounds) == np.inf):
```
I have a mind to change this too... for another PR. Mostly a note to self. This whole block could be made more concise.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1276,17 @@ def show_bounds(
+                cube_axes_actor.SetZAxisRange(axes_range[4], axes_range[5])
+            else:
+                raise ValueError(
+                    'axes_range must be passed as a [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
```
Depending on how you read it we might need "an" here: is it "_an xmin xmax..._" or is it "_a bracket xmin xmax ..._"?
```suggestion
                    'axes_range must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
```

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1276,17 @@ def show_bounds(
+            else:
+                raise ValueError(
+                    'axes_range must be passed as a [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
+                )
```
We should probably raise here `if axes_range is not None`, telling the user to do better.

adeak replied:
Github is being weird. In the "files" tab you can see the predecessor to this comment.

adeak replied:
Where is the exception raised when we pass in `axes_ranges = 'potato'`?

JacobBumgarner replied:
I see what you mean. I was assuming that the user wouldn't be passing anything absurd into the function.

Is this more agreeable?

```python

        # update the range values of each axis
        if axes_ranges is None:
            axes_ranges = bounds
        elif isinstance(axes_ranges, (collections.abc.Sequence, np.ndarray)):
            # Make sure the elements are all compatible 
            if not all(isinstance(i , (int, float)) for i in axes_ranges):
                raise TypeError('each element in axes ranges must be passed as an int or float')
            
            if len(axes_ranges) == 6:
                cube_axes_actor.SetXAxisRange(axes_ranges[0], axes_ranges[1])
                cube_axes_actor.SetYAxisRange(axes_ranges[2], axes_ranges[3])
                cube_axes_actor.SetZAxisRange(axes_ranges[4], axes_ranges[5])
            else:
                raise ValueError(
                    'axes_ranges must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
                )
```

adeak replied:
Not really, because this tackles the parentheses in my previous comment for which I said "let's ignore that for now" :P Not ignoring it is fine, but it's not what I had in mind originally. I'm talking about an `else` clause here!

As for the above snippet: it won't work for numpy array input. Instead I'd recommend converting `axes_ranges` to an array inside that `elif`, and checking the type. This is a bit fiddly which is why I suggested to ignore it for now. It's a larger issue if someone passes garbage input. "user passing something absurd into the function" is exactly that input validation should prepare for.

JacobBumgarner replied:
I interpreted your "let's ignore that for now" as, "let's ignore the fact that strings match the `is instance`". 

Honestly - this seems pedantic to me here... It seems excessive to write an error for every possible instance of potential argument that the user might pass. For example, if you pass in `plotter.show_bounds(bounds='potato')`, you only get a `numpy` error. 

Below fixes the np.array not passing correctly issue.
```python
if axes_ranges is None:
    axes_ranges = bounds
elif isinstance(axes_ranges, (collections.abc.Sequence, np.ndarray)):
    axes_ranges = np.asarray(axes_ranges)
    # make sure the elements are all compatible 
    if not isinstance(axes_ranges.dtype, (np.int_, np.float_)):
        raise TypeError('each element in axes_ranges must be passed as an int or float')

    if len(axes_ranges) == 6:
        cube_axes_actor.SetXAxisRange(axes_ranges[0], axes_ranges[1])
        cube_axes_actor.SetYAxisRange(axes_ranges[2], axes_ranges[3])
        cube_axes_actor.SetZAxisRange(axes_ranges[4], axes_ranges[5])
    else:
        raise ValueError(
            'axes_ranges must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
        )
```

If not this, how exactly are you hoping for me to catch non-confirming inputs? Thanks

JacobBumgarner replied:
Or this if not the former? This won't give a meaningful error to `axes_ranges=[0,1,0,1,0,'b']`, but seems a bit cleaner.

```python
if axes_ranges is None:
    axes_ranges = bounds
elif isinstance(axes_ranges, (list, tuple, np.ndarray)):
    if len(axes_ranges) == 6:
        cube_axes_actor.SetXAxisRange(axes_ranges[0], axes_ranges[1])
        cube_axes_actor.SetYAxisRange(axes_ranges[2], axes_ranges[3])
        cube_axes_actor.SetZAxisRange(axes_ranges[4], axes_ranges[5])
    else:
        raise ValueError(
            'axes_ranges must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
        )
else:
    raise TypeError('axes_ranges must be passed as a list, tuple, or np.ndarray with int or float elements')
```

adeak replied:
Yes, it's pedantic; I thought we'd met before :smile: But it's not pointless pedantry. Let me explain my stance.

If you do
```py
if cond1:
    foo  # handle default
elif cond2:
    bar  # handle sane inputs
baz  # rest of the code
```
you make the decision to silently ignore garbage input. That's not what Python usually does, and that's arguably not what a library (having downstream libraries as well as lots of users) should do. What I recommend is your last snippet:
```py
if cond1:
    foo  # handle default
elif cond2:
    bar  # handle sane inputs
else:
    raise TypeError("I don't think so.")
```

Of course nobody in their right mind would pass `{'potato'}`. If someone intentionally passes garbage they had it coming. But what _might_ happen is that someone passes the wrong variable by accident. Say, instead of passing `other_mesh.bounds` they pass `other_mesh`, which is not a sequence. In this case we would just silently ignore the `axes_ranges` kwarg, even though someone went out of their way to specify it.

Handling almost-valid inputs like `axes_ranges=[0,1,0,1,0,'b']` is not something we should always prepare for, at least not with too helpful error messages. This is a subtler issue, but an error elsewhere is an easier thing to track down than a silent no-op. We _could_ handle this case too with my earlier recommendation, but we don't really _have_ to; this would be something like
```py
elif isinstance(axes_ranges, (collections.abc.Sequence, np.ndarray):
    axes_ranges = np.asanyarray(axes_ranges)
    if not np.issubdtype(axes_ranges.dtype, np.number):
        raise TypeError(f'Input axes_ranges must be a numeric sequence.')
```
This would work because if there's a single string in there then the dtype would be `object`. And after this point we can just check that the shape of the range is `(6,)`. Again: we don't have to do this. I only feel strongly about my original point.

JacobBumgarner replied:
>Yes, it's pedantic; I thought we'd met before 😄 But it's not pointless pedantry. Let me explain my stance.

🤣 touché

>Of course nobody in their right mind would pass {'potato'}. If someone intentionally passes garbage they had it coming. But what might happen is that someone passes the wrong variable by accident. Say, instead of passing other_mesh.bounds they pass other_mesh, which is not a sequence. In this case we would just silently ignore the axes_ranges kwarg, even though someone went out of their way to specify it.

This scenario makes much more sense to me. I suppose I wasn't being creative enough to visualize it. :)

Based on your input, I'll go ahead and propose the following to try to deal with egregiously incorrect inputs as well as inputs that the user might not be aware of passing incorrectly...

```python
if axes_ranges is None:
    axes_ranges = bounds
elif isinstance(axes_ranges, (collections.abc.Sequence, np.ndarray)):
    axes_ranges = np.asarray(axes_ranges)
    if not isinstance(axes_ranges, np.number):
        raise TypeError(f'input axes_ranges must be a numeric sequence.')
    
    if axes_ranges.shape[0]:
        cube_axes_actor.SetXAxisRange(axes_ranges[0], axes_ranges[1])
        cube_axes_actor.SetYAxisRange(axes_ranges[2], axes_ranges[3])
        cube_axes_actor.SetZAxisRange(axes_ranges[4], axes_ranges[5])
    else:
        raise ValueError(
            'axes_ranges must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
        )
else:
    raise TypeError('input axes_ranges must be a numeric sequence')

```

JacobBumgarner replied:
Thanks for being patient enough to explain your thought process and some more of the design philosophy. I appreciate it :)

adeak replied:
> Thanks for being patient 

And the same to you ;)

> Based on your input, I'll go ahead and propose the following [...]

There are a few things that need editing in there:
1. `np.asanyarray` (note the "any") on line 4
2. take a closer look at my last snippet in my previous comment: `isinstance` -> `issubdtype` and using `axes_ranges.dtype`. Since we'll always have a numpy ndarray here, what we have to do is confirm that it has numbers inside.
3. `if axes_ranges.shape[0]` should be `if axes_range.shape == (6,)` to make sure we don't have something weird like a 2d array. (Or we could `ravel()` the array before working with it...)

You can also just go ahead and make the change and we can iron out the kinks as suggested changes on the PR.

adeak replied:
One style typo that I'll fix in a moment, otherwise the current version looks good to me, thanks.

adeak commented on tests/test_renderer.py:
```diff
@@ -1,10 +1,25 @@
 import pytest
 
 import pyvista
+from pyvista import _vtk
```
Almost all our tests use `import vtk`, treating the test as external to the library. I think we should do that here too (and probably fix up the other outliers eventually).
```suggestion
import vtk
```

adeak commented on tests/test_renderer.py:
```diff
@@ -1,10 +1,25 @@
+    bounds = (0.0, 1.0, 0.0, 1.0, 0.0, 1.0)  # Renderer bounds
+    axes_range = (0.0, 5.0, 0.0, 5.0, 0.0, 5.0)  # Axes bounds
+    plotter.show_bounds(bounds=bounds, axes_range=axes_range)
+    cube_axes_actor = plotter.renderer.cube_axes_actor  # Pull the actor
+    assert isinstance(cube_axes_actor, _vtk.vtkCubeAxesActor)  # Check type
+    assert cube_axes_actor.GetBounds() == bounds  # Check the bounds
+    # Check the axes ranges
+    assert cube_axes_actor.GetXAxisRange() == axes_range[0:2]
+    assert cube_axes_actor.GetYAxisRange() == axes_range[2:4]
+    assert cube_axes_actor.GetZAxisRange() == axes_range[4:]
```
Looking at this I realise that my suggestion to expose the `axes_range` as a property wouldn't easily work, because it's an attribute that lives on the cube axes actor. I had `bounds` in mind, but due to the padding even the bounds of the actor could be different from the bounds of the mesh. It might not be worth trying to expose that at the renderer level (assuming it would even make sense to define a single bounds and axes range for renderers).

JacobBumgarner replied:
See above for the added plotter property suggestion.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1276,17 @@ def show_bounds(
             raise ValueError(f'padding ({padding}) not understood. Must be float between 0 and 1')
         cube_axes_actor.SetBounds(bounds)
 
+        # update the values of range of each axis
```
```suggestion
        # update the range values of each axis
```

adeak reviewed (commented):
Thanks, looks good, just some minor nitpicks.

JacobBumgarner commented on pyvista/plotting/plotting.py:
```diff
@@ -1142,6 +1142,14 @@ def bounds(self):
 
+    @property
+    def axes_ranges(self):
+        """Return the axes ranges of the active renderer."""
```
In your opinion, does it make sense to have this property exposed in the main plotter?

adeak replied:
I would either expose it on the plotter, or nowhere. You typically don't look at `renderer` properties directly. Whether we should expose this: not sure. To me it makes sense to be to be able to query everything stateful in at least one way. Then again if a full `show_bounds()` call is the only thing that can set the axes ranges, then perhaps this is not that important... Let's leave this open for now and we'll see if anyone else chimes in.

JacobBumgarner commented on pyvista/plotting/renderer.py:
```diff
@@ -1248,7 +1254,10 @@ def show_bounds(
-            bounds = np.array(mesh.GetBounds())
+            bounds = np.array(mesh.bounds)
+        else:
+            bounds = np.asanyarray(bounds, dtype=np.float64)
```
Changed to `float`.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1276,27 @@ def show_bounds(
+        elif isinstance(axes_ranges, (collections.abc.Sequence, np.ndarray)):
+            axes_ranges = np.asanyarray(axes_ranges)
+            if not np.issubdtype(axes_ranges.dtype, np.number):
+                raise TypeError('input axes_ranges must be a numeric sequence.')
```
```suggestion
                raise TypeError('Input axes_ranges must be a numeric sequence.')
```

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1267,6 +1276,27 @@ def show_bounds(
+                    'axes_ranges must be passed as an [xmin, xmax, ymin, ymax, zmin, zmax] list or tuple'
+                )
+        else:
+            raise TypeError('input axes_ranges must be a numeric sequence')
```
```suggestion
            raise TypeError('Input axes_ranges must be a numeric sequence')
```

JacobBumgarner commented on pyvista/plotting/plotting.py:
```diff
@@ -1142,6 +1142,14 @@ def bounds(self):
+        """Return the axes ranges of the active renderer."""
+        if hasattr(self.renderer, 'cube_axes_actor'):
+            return self.renderer.axes_ranges
+        else:
+            return self.bounds
```
Removed.

JacobBumgarner commented:
Hi all, frankly I'm not sure why the unit testing is failing on the Linux machines. I can only run the testing locally on macOS/Windows, and those are passing. Any suggestions?

adeak commented:
Sorry, I won't have time for a few days to take a closer look at this. But I can confirm the breakage on linux with a large image regression error in `test_plotting.py::test_plot_show_grid`:

image cache
![plot of sphere with a grid in the background from ~-120 to 120 in range](https://user-images.githubusercontent.com/17914410/169902034-de307706-60e9-46a4-bb5e-3e1f084d5ed2.png)

this PR:
![plot of sphere with a grid in the background from -1 to 1 in range](https://user-images.githubusercontent.com/17914410/169902062-ea021f97-e04c-433c-802e-4fc592614039.png)

I'm surprised this is OS-specific (I checked and the test is not skipped on Windows), but anyway it seems that the grid bounds are normalised in this PR, which isn't a thing in the image cache (and on WIndows, it seems).

MatthewFlamm commented:
If you switch the `show_grid` and `add_mesh` lines it will pass:

https://github.com/pyvista/pyvista/blob/e731df426ebe2de144549329714dae52cf5b10ba/tests/plotting/test_plotting.py#L457-L461

It seems like before this PR the bounds get updated automatically when a new mesh is added after `show_grid` is called, but with this PR the bounds do not get automatically updated even when no `axes_range` is provided.  This seems like a regression to me.

JacobBumgarner commented:
@MatthewFlamm Ok - I believe I've found the cause of the problem. Honestly, I'm not sure why the regression testing is passing on my macbook! It shouldn't be.

Below is the final line of the `plotter.add_mesh(...)` function.

https://github.com/pyvista/pyvista/blob/55917fa2dcfa3ad62b46619f39637908b9dc6a18/pyvista/plotting/plotting.py#L2401

[Here is the VTK doc](https://vtk.org/doc/nightly/html/classvtkObject.html#a8586479b9d204cf0a45fb693d4281354) of the `vtkObject.Modified()` function, which is a super of the `vtkRenderer`

It turns out that when the axes ranges are explicitly set with the `vtkCubeAxesActor.Set_AxisRange(...)`,  the `renderer.Modified()` call doesn't update the axes with the bounds of the added mesh(es).

---

To resolve this, perhaps unless the axes_ranges are explicitly passed, we won't call `vtkCubeAxesActor.Set_AxisRange(...)`? I.e., we only expose the call to `Set_AxisRange(...)` if users specifically want to set the ranges. Otherwise, the axes ranges will be set internally by VTK and will be updated when we call `renderer.Modified()`.

Does this approach seem reasonable to you?

JacobBumgarner commented:
Actually, is this the reason that the regression testing is passing on macOS/windows?

https://github.com/pyvista/pyvista/blob/eaa71a640264765c8a9b40d485b6ccdb970c87b0/tests/plotting/test_plotting.py#L111-L113

adeak commented:
Absolutely, good catch. Also good to know about that...

MatthewFlamm commented:
> Actually, is this the reason that the regression testing is passing on macOS/windows?

Yes.  Not sure if this could be removed, or maybe have a flag that it could be turned on or off.

> To resolve this, perhaps unless the axes_ranges are explicitly passed, we won't call vtkCubeAxesActor.Set_AxisRange(...)? I.e., we only expose the call to Set_AxisRange(...) if users specifically want to set the ranges. Otherwise, the axes ranges will be set internally by VTK and will be updated when we call renderer.Modified()

This approach makes sense to me.

akaszynski commented:
> I might need to create a separate image cache for each OS, which may be non-trivial.

Turns out this isn't that bad. Implemented in #2716.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1263,7 +1270,10 @@ def show_bounds(
 
         # set bounds
         if bounds is None:
-            bounds = np.array(mesh.GetBounds())
+            bounds = np.array(mesh.bounds)
+        else:
+            bounds = np.asanyarray(bounds, dtype=float)
```
This looks good to me, but perhaps it would be easier to follow the execution path if this is moved to line 1220 where a default bounds is already defined.

akaszynski replied:
Done in 27624437a5b9f6e0e8112b3aaed57200bbb6de84.

adeak reviewed (dismissed):
Sorry for dragging my feet with this review. This looks good to me with one minor suggestion for maintainability. Approving anyway.

We'll also have to check coverage once we unbreak CI :)

akaszynski commented:
> It seems like before this PR the bounds (or axes ranges?) get updated automatically when a new mesh is added after `show_grid` is called, but with this PR the bounds (or axes range?) do not get automatically updated even when no `axes_range` is provided. This seems like a regression to me.

I've addressed this in 27624437a5b9f6e0e8112b3aaed57200bbb6de84. The new behavior is such that when setting bounds, they remain fixed for the cube axes actor (expected behavior), and when bounds are not set, the axes actor bounds are fluid (also expected).

---

Added an additional test to improve coverage and cleaned up the errors a bit. I really prefer it when errors tell me what I'm supposed to rather than just telling me I did it wrong.

akaszynski commented:
> This looks nice, @JacobBumgarner! Would you be able to add a trivial test for code coverage?
> 
> Also, I'm curious if you could provide a screenshot (perhaps two subplots) showing this feature and how it improves the bounds actor for the types of data you mentioned?

Addressed in f3dcf288fd3bbf9f1a4eb3ddf6a5ef2b3b332e25.

adeak commented on tests/plotting/test_plotting.py:
```diff
@@ -521,11 +521,31 @@ def test_plot_bounds_axes_with_no_data():
     plotter.show(before_close_callback=verify_cache_image)
 
 
+def test_plot_show_grid_with_mesh(hexbeam, plane):
+    """Show the grid bounds for a specific mesh."""
+    hexbeam.clear_data()
+    plotter = pyvista.Plotter()
+    plotter.add_mesh(hexbeam, style='wireframe')
+    plotter.add_mesh(plane)
+    plotter.show_grid(mesh=plane)
+    plotter.show(before_close_callback=verify_cache_image)
```
What's up with that `Z axis (x10^-18)` annotation? Is it applying the bounds to the plane, which being 2d has 0 Z range? Should we switch the order of the plane and the beam being added just to make this a bit nicer?

akaszynski replied:
Might be better to just disable showing the z axis labels. If this was for an example I’d agree with changing it, but if it’s just for a unit test, there’s not much of a point in making it look nice.

adeak replied:
It might also happen that the `1e-18` noise ends up changing between systems or VTK versions, but disabling it altogether will prevent that too.

akaszynski replied:
Good point. I’ll disable it.

MatthewFlamm replied:
You could also set the label format to be not in scientific format.

akaszynski replied:
Done in 366e4f67152d08735e46748f81eecc4b634fd1.

JacobBumgarner commented:
Sorry to all for letting this one slip off my radar, things have been quite hectic over here.

akaszynski reviewed (approved):
This LGTM, but would appreciate a second review.

akaszynski commented:
> Sorry to all for letting this one slip off my radar, things have been quite hectic over here.

No problem at all! These PRs can take a while and sometimes it takes a little help to get it over the last bit.

adeak commented on examples/02-plot/bounds.py:
```diff
@@ -0,0 +1,72 @@
+
+gears = examples.download_gears()
+
+plotter = pv.Plotter()
+plotter.add_mesh(gears, smooth_shading=True, split_sharp_edges=True)
+plotter.show_bounds(axes_ranges=[0, 5, 0, 5, 0, 2], show_zlabels=False)
+plotter.show()
+
```
I assume this is a VTK thing, but this is what the plot looks like:
![Screenshot from 2022-06-18 15-24-53](https://user-images.githubusercontent.com/17914410/174439732-53d12cd7-f7fc-4f34-b17e-67a4db8a6089.png)
note the arrows on the axis lines pointing from higher values toward lower ones.

akaszynski replied:
I think this is a VTK “feature”. Reversing the range doesn’t work either (e.g. manually providing reversed bounds).

adeak replied:
Nevermind, I'm dumb. Those are not arrows, those are ticks plotted in the xy/xz/yz planes... they just look bad.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1259,11 +1281,13 @@ def show_bounds(
             else:
-                raise ValueError(f'Value of location ({location}) not understood.')
+                raise ValueError(
+                    f'Value of location ("{location}") should be either "all", "origin",'
+                    ' "outer", "default", "closest", "front", "furthest", or "back".'
```
This list includes `'default'`, `'closest'`, `'furthest'` but these aren't mentioned in the docstring.

akaszynski replied:
Fixed in 04db34ed982fe7647212ca4233939c869be538ad.

adeak commented on pyvista/plotting/renderer.py:
```diff
@@ -1047,6 +1047,7 @@ def show_bounds(
         self,
         mesh=None,
         bounds=None,
+        axes_ranges=None,
```
I noticed that `axes_ranges` is no longer an attribute on the `Plotter`. I assume this is on purpose, and I assume because `renderer.bounds` probably also doesn't reflect custom `bounds` being passed to `show_bounds()`. But it's better to discuss this change explicitly.

akaszynski replied:
That will have to be a plotter property if we add it, so we can document it. Do you think we need this?

adeak replied:
I remembered discussing this earlier, in https://github.com/pyvista/pyvista/pull/2441#discussion_r847667283. But apparently @JacobBumgarner ended up thinking that it's not worth having in the `Plotter` API, and I'd be on the fence myself, so I think this can stay as is if you agree.

JacobBumgarner replied:
Yes that's correct.  I decided that adding the `axes_ranges` attribute to the plotter would only bloat it, as I image the use case is quite small. 

Moreover, the property is only relevant when a vtkCubeAxesActor was added to the plotter.

adeak reviewed (commented):
Just some minor remarks and a small change pushed in a moment.

adeak reviewed (approved):
Thanks, LGTM!
